File: Python/DayOfWeek/date.py
months = {
    "Jan": 1,
    "Feb": 2,
    "Mar": 3,
    "Apr": 4,
    "May": 5,
    "Jun": 6,
    "Jul": 7,
    "Aug": 8,
    "Sep": 9,
    "Oct": 10,
    "Nov": 11,
    "Dec": 12
}
monthDays = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
weekdays = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"]

# Returns the number of leap years from year 0 to year yr
def countLeapYears(mon, yr):
    if months[mon] <= 2:
        yr -= 1
    return int(yr / 4) - int(yr / 100) + int(yr / 400)

# Returns the total number of days at date "mon/day/yr" since Dec 31 1BC 
def countDays(day, mon, yr):
    # Set total_days to the number of days elapsed in the current month
    total_days = day

    # Iterate through monthDays and add total number of days in months already elapsed
    for m in range(0, months[mon] - 1):
        total_days += monthDays[m]
    
    # Add 365 days for every year and an extra day for every leap year
    total_days += (365 * yr) + countLeapYears(mon, yr)

    return total_days

# Returns the number of days between dates "mon1/day1/yr1" and "mon2/day2/yr2"
def getDayDiff(day1, mon1, yr1, day2, mon2, yr2):
    return countDays(day2, mon2, yr2) - countDays(day1, mon1, yr1)

# Returns the days of the week of date "mon/day/yr"
def getDayOfWeek(day, mon, yr):
    day_number = getDayDiff(21, "Feb", 2021, day, mon, yr) % 7
    return weekdays[day_number]


print(getDayOfWeek(7, "Dec", 1941))
print(getDayOfWeek(6, "Jan", 2021))
print(getDayOfWeek(4, "Jul", 1776))

File: Java/Eshan/Date/Date.java
public class Date {
    int day;
    int month;
    int year;

    public Date(int day, int month, int year) {
        this.day = day;
        this.month = month;
        this.year = year;
    }

    public boolean isLeapYear() {
        return year % 400 == 0 || (year % 4 == 0 && year % 100 != 0);
    }

    public int getDaysInYear() {
        return isLeapYear() ? 366 : 365;
    }

    public int getDaysInMonth() {
        if (month == 1 || month == 3 || month == 5|| month == 7 || month == 8 || month == 10 || month == 12) {
            return 31;
        }
        if (month == 2) {
            return isLeapYear() ? 29 : 28;
        }
        return 30;
    }

    public void addDays(int days) {
        while (days + day > getDaysInMonth()) {
            days -= (getDaysInMonth() - day + 1);
            day = 1;
            if (month < 12) {
                month++;
            }
            else {
                month = 1;
                year++;
            }
        }
        day += days;
    }

    public int countDays() {
        Date current = new Date(0,1,0);
        int totalDays = day;
        while (current.year < year) {
            totalDays += current.getDaysInYear();
            current.year++;
        }

        while (current.month < month) {
            totalDays += current.getDaysInMonth();
            current.month++;
        }
        return totalDays;
    }
    

    public int daysTo(Date other) {
        return other.countDays() - countDays();
    }

    public String toString() {
        return String.format("%d/%d/%d", month, day, year);
    }
}
